server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+all for the UpdateResponse, preferring instead to poll for it later.
+
+### Invalid Update
+
+UpdateRequests can be invalid given the state of the receiving workflow. In such
+cases the UpdateResponse will contain an error indication and payload. Invalid
+updates _must not_ be recorded in workflow history to prevent a malicious issuer
+from causing a denial of service on the workflow by issuing repeated invalid
+UpdateRequests.
+
+### "Failing" Update
+
+UpdateRequests can be semantically acceptable but still fail due to external
+dependencies. In such cases the UpdateResponse will contain an error indication
+and payload.
+
+### Unexpected Disconnect
+
+While the issuer of an update may wish to block on an UpdateResponse, network
+conditions or software failures may occur which forcibly disconnects the update
+issuer from the server at the time that the Update completes _and_ before the
+issuer becomes aware of the Update ID. The issuer must be able to re-issue the
+same UpdateRequest to access the server-generated Update ID.
+
+## gRPC API
+
+The gRPC API comprises two functions, `UpdateWorkflow` and
+`AwaitUpdateWorkflowResponse` to be added to the `WorkflowService` service
+definition. The former is used to invoke an update and follows the standard
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc AwaitUpdateWorkflowResponse(AwaitUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
+}
+
+message UpdateWorkflowResponse {
+    UpdateResultType result_type = 1;
+    string update_id = 2;
+    temporal.api.common.v1.Payloads output = 3;
+}
+
+enum UpdateResultType {
+    UPDATE_RESULT_TYPE_UNSPECIFIED = 0;
```

Comment:
We never use the `UNSPECIFIED` variant in other API enums, we should have a distinct value for update_id.

Reply:
I see it used in plenty of places at https://github.com/temporalio/api/tree/master/temporal/api/enums/v1. I appreciate it due to following https://google.aip.dev/126, but in general I can see value in removing this altogether and just relying on presence of an error to determine errored.

Reply:
We have the value in all our enum definitions but we never use it, it means what it says: UNSPECIFIED

Reply:
Added UPDATE_WORKFLOW_RESULT_TYPE_IN_PROGRESS to avoid overloading UNSPECIFIED.